### PR TITLE
Sandbox: run start function.

### PR DESCRIPTION
**Pull request by pepyakin**
This changes allows the execution of the start function (if any).

### CODE REVIEW

File: substrate/executor/src/sandbox.rs [hunk 1]
```diff
@@ -279,12 +263,50 @@ impl<'a, FE: SandboxCapabilities + Externals + 'a> Externals for GuestExternals<
...
+}
+
+pub struct SandboxInstance {
 	instance: ModuleRef,
 	dispatch_thunk: FuncRef,
 	guest_to_supervisor_mapping: GuestToSuperviserFunctionMapping,
 }
 
+impl SandboxInstance {
+	pub fn invoke<FE: SandboxCapabilities + Externals>(
```
Comment (rphmeier) [inline substrate/executor/src/sandbox.rs]: doc?

Comment (pepyakin) [reply substrate/executor/src/sandbox.rs]: Done!

Also added https://github.com/paritytech/polkadot/pull/153/commits/47b6a9f537484e224d77d3a0ec9225b9fb9cdc5a to the module

**Comment by gavofyork** [comment]
what is the `start` function meant to do?

**Comment by pepyakin** [comment]
The `start` function is run at the very last step of a module instantiation process. It's needed to do last time initialization, that isn't doable with memory data segments (e.g. calling a syscall).

One of the use-cases might be C++ global initialization routines which are called before execution of `main`. However, when I did experiments I was unable to make clang to emit `start` function, though I didn't try hard enough.

Regarding Rust: as far as I know, it doesn't use such global initialization and doesn't emit `start` function.

While the `start` function might seem somewhat useless to us at the moment, I prefer to deal with it now since effects of not running `start` function if and when it is actually emited might lead to subtle breakages. Also, there is no way to skip running the `start` function in JS.

**Comment by gavofyork** [comment]
fair enough.

**Pull request merged**